### PR TITLE
nautilus: os/bluestore: support RocksDB prefetch in buffered read mode

### PR DESCRIPTION
**Pull request by ifed01**
Backports: https://github.com/ceph/ceph/pull/27782
No tracker.

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -1448,9 +1451,12 @@ int BlueFS::_read(
     outbl->clear();
 
   int ret = 0;
+  std::shared_lock s_lock(h->lock);
   while (len > 0) {
     size_t left;
     if (off < buf->bl_off || off >= buf->get_buf_end()) {
+      s_lock.unlock();
+      std::unique_lock u_lock(h->lock);
```
Comment (liewegas) [inline src/os/bluestore/BlueFS.cc]: this looks wrong.  if we drop and retake the lock, don't we need to revalidate the conditions checked above?

Comment (ifed01) [reply src/os/bluestore/BlueFS.cc]: You're right. See https://github.com/ceph/ceph/pull/29012 for the fix.

Comment (ifed01) [reply src/os/bluestore/BlueFS.cc]: @liewegas - fix has been cherry-picked from #29012

**Comment by yuriw** [comment]
[wip-yuri4-testing-2019-07-24-2314-nautilus](https://trello.com/c/QnlxiPxf/675-wip-yuri4-testing-2019-07-24-2314-nautilus)

**Comment by yuriw** [comment]
test this please

**Pull request merged**